File: shubhmjain30/1-basic/pattern/pattern.js
// questions: https://takeuforward.org/strivers-a2z-dsa-course/must-do-pattern-problems-before-starting-dsa/

const pattern_1 = (count) => {
	let res;
	for (let i = 1; i < count + 1; i++) {
		res = "";
		for (let j = 1; j < count + 1; j++) {
			res += "*";
		}
		console.log(res, "\n");
	}
};

const pattern_2 = (count) => {
	let res;
	for (let i = 1; i < count + 1; i++) {
		res = "";
		for (let j = 1; j < i + 1; j++) {
			res += "*";
		}
		console.log(res, "\n");
	}
};

const pattern_3 = (count) => {
	let res;
	for (let i = 1; i < count + 1; i++) {
		res = "";
		for (let j = 1; j < i + 1; j++) {
			res += `${j}`;
		}
		console.log(res, "\n");
	}
};

const pattern_4 = (count) => {
	let res;
	for (let i = 1; i < count + 1; i++) {
		res = "";
		for (let j = 1; j < i + 1; j++) {
			res += `${i}`;
		}
		console.log(res, "\n");
	}
};

const pattern_5 = (count) => {
	let res;
	for (let i = count; i > 0; i--) {
		res = "";
		for (let j = i; j > 0; j--) {
			res += "*";
		}
		console.log(res, "\n");
	}
};

const pattern_6 = (count) => {
	let res;
	for (let i = count; i > 0; i--) {
		res = "";
		for (let j = i; j > 0; j--) {
			res += `${i - j + 1}`;
		}
		console.log(res, "\n");
	}
};

const pattern_7 = (count) => {
	let res;
	let col = count * 2 - 1;
	let mid = Math.floor(col / 2);
	for (let i = 0; i < count; i++) {
		res = "";
		start = mid - i;
		end = mid + i;
		for (let j = 0; j < col; j++) {
			if (j >= start && j <= end) {
				res += `*`;
			} else {
				res += `${j}`;
			}
		}
		console.log(res, "\n");
	}
};

const pattern_8 = (count) => {
	let res;
	let col = count * 2 - 1;
	for (let i = 0; i < count; i++) {
		res = "";
		let start = i;
		let end = col - i - 1;
		for (let j = 0; j < col; j++) {
			if (j >= start && j <= end) {
				res += "*";
			} else {
				res += " ";
			}
		}
		console.log(res, "\n");
	}
};

(() => {
	const func_list = [
		pattern_1,
		pattern_2,
		pattern_3,
		pattern_4,
		pattern_5,
		pattern_6,
		pattern_7,
		pattern_8,
	];
	const default_count = 5;
	const count_list = [3, 5, 5, 6, 5, 8, 4, 5];

	for (let i = 0; i < func_list.length; i++) {
		console.log(
			`\n\n\n>>>>>> RUNNING ${func_list[i].name} with argument ${
				count_list[i] ?? default_count
			} >>>>>> \n`
		);
		func_list[i](count_list[i] ?? default_count);
	}
})();

File: praj-tarun/1_The_Basics/pattern/pattern.py
def pattern_1(n):
    i=0
    while i<n:
        j=0
        while j<n:
            print('*',end='')
            j+=1
        print()
        i+=1

def pattern_2(n):
    i=0
    while i<n:
        j=n-i
        while j<=n:
            print('*',end='')
            j+=1
        print()
        i+=1

def pattern_3(n):
    i=1
    while i<=n:
        j=1
        while j<=i:
            print(j, end='')
            j+=1
        print()
        i+=1
        
def pattern_4(n):
    i=1
    while i<=n:
        j=1
        while j<=i:
            print(i, end='')
            j+=1
        print()
        i+=1

def pattern_5(n):
    i=n
    while i>0:
        j=i
        while j>0:
            print('*',end='')
            j-=1
        print()
        i-=1

def pattern_6(n):
    i=1
    while i<=n:
        j=1
        while j<=n-i+1:
            print(j, end='')
            j+=1
        print()
        i+=1

def pattern_7(n):
    i=0
    while i<n:
        j=1
        while j<=n-i:
            print(' ', end='')
            j+=1
        j=1
        while j<=2*i+1:
            print('*',end='')
            j+=1
        j=1
        while j<=n-i:
            print(' ', end='')
            j+=1
        print()
        i+=1

def pattern_8(n):
    i=n-1
    while i>=0:
        j=1
        while j<=n-i:
            print(' ', end='')
            j+=1
        j=1
        while j<=2*i+1:
            print('*',end='')
            j+=1
        j=1
        while j<=n-i:
            print(' ', end='')
            j+=1
        print()
        i-=1

def pattern_9(n):
    pattern_7(n)
    pattern_8(n)

def pattern_10(n):
    pattern_2(n)
    pattern_5(n-1)

def pattern_11(n):
    i=1
    while i<=n:
        j=1
        while j<=i:
            if (i+j)%2==0:
                print(1,end='')
            else:
                print(0,end='')
            j+=1
        print()
        i+=1

def pattern_12(n):
    i=1
    while i<=n:
        j=1
        while j<=n*2:
            if j<=i:
                print(j,end='')
            elif j>2*n-i:
                print(2*n-j+1,end='')
            else:
                print(' ',end='')
            j+=1
        print()
        i+=1

def ispalindrome(n):
    l=0
    r=len(n)-1
    while l<r:
         
        

def main(): 
    pattern_12(4)
 


if __name__ == "__main__":
    main()